## Exercise 3 -- dating Neanderthal admixture

(This exercise is a bonus for those of you who are already experts in R. If you're not very comfortable with R or population genetics, don't feel like you have to rush through the exercises to finish everything. You can take a look at this exercise at home, if you want.)

### Introduction

*You sequenced 100 diploid genomes from a Eurasian population and are interested in estimating the time of Neanderthal introgression into the ancestors of this population. The literature suggests that the introgression happened around 55 thousand years ago -- does this also apply to the population that you sequenced?*

*To be able to do this, you ran an inference software which gives you the exact coordinates of Neanderthal DNA tracts present in every Eurasian genome that you sequenced. This of course means that you also know the lengths of each of those tracts.*

*Use the distribution of the Neanderthal tract lengths in your population to estimate the time of Neanderthal introgression!*

### Moving over to R

Type `R` in your terminal or (better) just use RStudio R console if you have it.

#### Task: Load and inspect the tracts data

**First load the table with coordinates of all Neanderthal tracts** into R:

```{r}
tracts <- read.table(url("https://github.com/bodkan/ku-introgression2025/raw/main/tracts.tsv"), sep = "\t", header = TRUE)
```

**Familiarize yourself with the data** by running this R command which shows information from only the first few genotypes:

```{r}
head(tracts)
```

**For how many individuals do we have information about Neanderthal DNA tracts that they carry?**

```{r}
length(unique(tracts$individual))
```

#### Task: plot the distribution of tract lengths across bins

It looks like the inference software (or a helpful bioinformatician) binned each tract according to its length (see the column `bin`). **What does the distribution of Neanderthal tract lengths looks like in your data?** Knowing that recombination has acted on the introgressed Neanderthal DNA over time, each generation, suggests that the distribution should look exponential -- do you see this in the data? To answer this, plot the proportion of tracts in each bin.

```{r tract_bins}
# get the bin numbers
bins <- sort(unique(tracts$bin))

# count the tracts in each bin and compute the proportion of tracts in each bin
counts <- as.integer(table(tracts$bin))
props <- counts / sum(counts)

plot(bins, props, xlab = "Neanderthal tract length bin", ylab = "proportion of tracts")
```

The distribution does, indeed, look quite exponential. **Next we will try to use this to date Neanderthal introgression using the information encoded in the distribution of tract lengths!**

#### Dating the introgression event -- a bit of theory first

As we know, over time since admixture, recombination breaks up longer haplotypes into shorter ones, regularly almost like a clock. And it turns out that the distribution of tract lengths after time $t$ follows exponential decay, leading to the distribution of tract lengths $x$ to have the following form:

$$
\textrm{tract length of }~x \sim \exp(-\lambda x) = \exp(-r t x),
$$

where the $\lambda$ parameter determines the *rate* of exponential decay which can be computed, under some simplifying assumptions, as the product of the recombination rate (traditionally in humans with value of about $1 \times 10^{-8}$) and $t$ which is the time since admixture:

$$
\lambda = rt
$$

**The** $t$ in this equation latter is our unknown we're trying to compute in this exercise! So we know which equation we can use to extimate the admixture time.

It also turns out that the expected value of this exponential distribution plotted above (which [can be computed](https://en.wikipedia.org/wiki/Exponential_distribution#Mean,_variance,_moments,_and_median) simply as $1 / \lambda = 1 / rt$) gives us the theoretical expression for the expected tract length after time $t$.

#### Task: Compute the average length of an introgressed tract

Of course, you can also compute this expectation from the data, **so do this now: get an estimate of the expected length of an introgressed fragment after (unknown) time** $t$ by computing the average introgressed tract length observed in data. What is the average length of a Neanderthal DNA segment in the data? (Remember that at the moment of introgression, *entire Neanderthal chromosomes* were segregating in modern humans!)

```{r}
L <- mean(tracts$length)
L # length in units of base pairs
```

#### Task: date the introgression event -- in practice

Taking the simple math above together and doing a little algebra, we can express the average expected length of an introgressed fragment after time $t$ using this formula:

$$
\textrm{average tract length}~L = \frac{1}{\lambda} = \frac{1}{rt}
$$

**Because we know** $L$ as computed just above (average tract length as you just computed) and $r$ (recombination rate of $1 \times 10^{-8}$), this means we can estimate the time since the admixture by a simple rearrangement of the above equation to separate the time of introgression $t$:

$$
t = \frac{1}{rL},
$$

where `r` is the recombination rate and `L` is the average length of an introgressed Neanderthal tract (as you just computed it).

Note that this time estimate will be in units of generations, so we'll have to multiply this quantity by the generation time (roughly 30 years for humans) to get time in years before the present.

**Use this simple equation** $\frac{1}{rL}$ to compute the estimate of the admixture time:

```{r}
r <- 1e-8 # crossovers per bp per generation
L <- mean(tracts$length) # average tract length after time t

t <- 1 / (r * L)
t

# convert the time of introgression to years before present assuming
# generation time of 30 years
t * 30
```

You should get a value will be quite close to \~55 thousand years ago, an estimate which is often found in the literature as the time when Neanderthals and anatomically modern humans interbred!

------------------------------------------------------------------------

## Bonus content

As a last sanity check, if we use this time to compute the rate of exponential decay $\lambda$, we should get a nice fit of the theoretical exponential decay curve over the empirical counts of tract lengths in each bin. As a reminder, this is the decay:

```{r tract_bins2}
plot(bins, props, xlab = "Neanderthal tract length bin", ylab = "proportion of tracts")
```

Let's try if we can plot the theoretical exponential decay from the estimated time of admixture.

First, because the exponential tract decay you plotted above shows the decay of the length of introgressed tracts in bins, the $\lambda$ parameter determines the decay expected for tracts falling into to successive bins (so decay of longer and longer tracts). However, our recombination rate $r$ which features in the equation to compute $\lambda$ above describes recombination in units of base pairs, not bins. In order to be able to plot the theoretical exponential decay curve, we therefore first need to compute the average increase in tract lengths from bin to bin:

```{r}
# compute the average length in each bin
average_bins <- aggregate(length ~ bin, data = tracts, FUN = mean)

# compute the difference between bins to get "average bin step size"
bin_step <- mean(diff(average_bins$length))

bin_step
```

With this, we can overlay the theoretical exponential decay expectation across the empirical distribution of tract lengths in different bins:

```{r tract_fit}
r <- 1e-8 # crossovers per bp per generation
t <- 1800 # time of admixture (in generations) we computed above

# plot the empirical proportions of tracts in each bin
plot(bins, props, xlab = "Neanderthal tract length bin", ylab = "proportion of tracts")

# overlay theoretical exponential density curve assuming recombination rate r
# and time of introgression t generations ago
lambda <- r * bin_step * t
y <- dexp(bins, rate = lambda)
lines(bins, y, col = "red", lty = 2)
```

------------------------------------------------------------------------

If you want to take a closer look at how the tracts data was prepared (it was simulated!), you can see the complete code [here](generate_tracts.R).